types/wlr_screencopy_v1.c

```diff
@@ -10,8 +10,82 @@
 
 #define SCREENCOPY_MANAGER_VERSION 1
 
+struct screencopy_damage {
+	struct wl_list link;
+	struct wlr_output *output;
+	struct pixman_region32 damage;
+	struct wl_listener output_precommit;
+	bool is_handled;
+};
+
 static const struct zwlr_screencopy_frame_v1_interface frame_impl;
 
+static struct screencopy_damage *find_output_damage(
+		struct wlr_screencopy_manager_v1 *manager,
+		struct wlr_output *output) {
+	struct screencopy_damage *damage;
+
+	wl_list_for_each(damage, &manager->damages, link) {
+		if (damage->output == output) {
+			return damage;
+		}
+	}
+
+	return NULL;
+}
+
+static void accumulate_damage(struct screencopy_damage *damage) {
+	struct pixman_region32 *region = &damage->damage;
+	struct wlr_output *output = damage->output;
+
+	pixman_region32_union(region, region, &output->pending.damage);
+	pixman_region32_intersect_rect(region, region, 0, 0, output->width,
+		output->height);
+}
+
+static void damage_handle_output_precommit(struct wl_listener *listener,
+		void *data) {
+	struct screencopy_damage *damage =
+		wl_container_of(listener, damage, output_precommit);
+	if (!damage->is_handled) {
+		accumulate_damage(damage);
+	}
+}
+
+static struct screencopy_damage *new_output_damage(
```

Comment:
One `screencopy_damage` is created when a client starts capturing, but isn't free'ed until the compositor shuts down. This leaks resources.

Reply:
It is freed `in wlr_screencopy_manager_v1_destroy()`. Isn't that called when the client destroys its manager object or disconnects?
This is used for the per-client damage accumulation.

Reply:
No, `wlr_screencopy_manager` is not per-client. It's created once by the compositor for its whole lifetime.

Reply:
Ahh, right, it needs to go into `manager_handle_resource_destroy()`.